diff --git a/makefile.inc b/makefile.inc
index f2d93bf..c7dbffa 100644
--- a/makefile.inc
+++ b/makefile.inc
@@ -24,8 +24,8 @@ endif
 
 # -O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -fPIC
 
-CFLAGS      := $(CFLAGS) -Wall
-LDLIBS      := $(LDLIBS) -pthread
+CFLAGS      := $(CFLAGS) -Wall -fprofile-arcs -ftest-coverage
+LDLIBS      := $(LDLIBS) -pthread -lgcov
 LDMYSQL     := -lmysqld
 CPP_LDLIBS  := -lstdc++
 
diff --git a/server/core/Makefile b/server/core/Makefile
index cb0250d..fe0f579 100644
--- a/server/core/Makefile
+++ b/server/core/Makefile
@@ -46,7 +46,7 @@ CC=cc
 CFLAGS=-c -I/usr/include -I../include -I../modules/include -I../inih \
 	$(MYSQL_HEADERS) \
 	-I$(LOGPATH) -I$(UTILSPATH) \
-	-Wall -g
+	-Wall -g -fprofile-arcs -ftest-coverage
 
 include ../../makefile.inc
 
@@ -76,7 +76,7 @@ POBJS=maxpasswd.o secrets.o utils.o
 LIBS=-L$(EMBEDDED_LIB) \
 	-lmysqld \
 	-lz -lm -lcrypt -lcrypto -ldl -laio -lrt -pthread -llog_manager \
-	-L../inih/extra -linih -lssl -lstdc++ 
+	-L../inih/extra -linih -lssl -lstdc++ -lgcov
 
 all:	maxscale maxkeys maxpasswd
 
diff --git a/server/modules/filter/Makefile b/server/modules/filter/Makefile
index f8cf910..ca022b8 100644
--- a/server/modules/filter/Makefile
+++ b/server/modules/filter/Makefile
@@ -25,7 +25,7 @@ UTILSPATH := $(ROOT_PATH)/utils
 
 CC=cc
 CFLAGS=-c -fPIC -I/usr/include -I../include -I../../include -I$(LOGPATH) \
-	-I$(UTILSPATH) -Wall -g
+	-I$(UTILSPATH) -Wall -g -fprofile-arcs -ftest-coverage
 
 include ../../../makefile.inc
 
@@ -44,7 +44,7 @@ TEESRCS=tee.c
 TEEOBJ=$(TEESRCS:.c=.o)
 SRCS=$(TESTSRCS) $(QLASRCS) $(REGEXSRCS) $(TOPNSRCS) $(TEESRCS)
 OBJ=$(SRCS:.c=.o)
-LIBS=$(UTILSPATH)/skygw_utils.o -lssl -llog_manager
+LIBS=$(UTILSPATH)/skygw_utils.o -lssl -llog_manager -lgcov
 MODULES= libtestfilter.so libqlafilter.so libregexfilter.so libtopfilter.so libtee.so libhintfilter.so
 
 
diff --git a/server/modules/filter/hint/Makefile b/server/modules/filter/hint/Makefile
index 4f21947..3ad0e3b 100644
--- a/server/modules/filter/hint/Makefile
+++ b/server/modules/filter/hint/Makefile
@@ -25,7 +25,7 @@ UTILSPATH := $(ROOT_PATH)/utils
 
 CC=cc
 CFLAGS=-c -fPIC -I/usr/include -I../../include -I../../../include -I$(LOGPATH) \
-	-I$(UTILSPATH) -Wall -g
+	-I$(UTILSPATH) -Wall -g -fprofile-arcs -ftest-coverage
 
 include ../../../../makefile.inc
 
@@ -34,7 +34,7 @@ LDFLAGS=-shared  -L$(LOGPATH) -Wl,-rpath,$(DEST)/lib \
 
 SRCS= hintfilter.c hintparser.c
 OBJ=$(SRCS:.c=.o)
-LIBS=$(UTILSPATH)/skygw_utils.o -lssl -llog_manager
+LIBS=$(UTILSPATH)/skygw_utils.o -lssl -llog_manager -lgcov
 
 libhintfilter.so: $(OBJ)
 	$(CC) $(LDFLAGS) $(OBJ) $(LIBS) -o $@
diff --git a/server/modules/monitor/Makefile b/server/modules/monitor/Makefile
index f815efe..308f496 100644
--- a/server/modules/monitor/Makefile
+++ b/server/modules/monitor/Makefile
@@ -25,7 +25,7 @@ UTILSPATH := $(ROOT_PATH)/utils
 
 CC=cc
 CFLAGS=-c -fPIC -I. -I/usr/include -I../include -I../../include -I$(LOGPATH) \
-	-I$(UTILSPATH) $(MYSQL_HEADERS) -Wall -g
+	-I$(UTILSPATH) $(MYSQL_HEADERS) -Wall -g -fprofile-arcs -ftest-coverage
 
 LDFLAGS=-shared  -L$(LOGPATH) -Wl,-rpath,$(DEST)/lib \
 	-Wl,-rpath,$(LOGPATH) -Wl,-rpath,$(UTILSPATH) \
@@ -42,7 +42,7 @@ NDBCLUSTEROBJ=$(NDBCLUSTERSRCS:.c=.o)
 SRCS=$(MYSQLSRCS) $(GALERASRCS) $(NDBCLUSTERSRCS)
 OBJ=$(SRCS:.c=.o)
 LIBS=$(UTILSPATH)/skygw_utils.o -llog_manager \
-	-L$(EMBEDDED_LIB) -lmysqld
+	-L$(EMBEDDED_LIB) -lmysqld -lgcov
 MODULES=libmysqlmon.so libgaleramon.so libndbclustermon.so
 
 
diff --git a/server/modules/protocol/Makefile b/server/modules/protocol/Makefile
index 54a8f8c..c8913ab 100644
--- a/server/modules/protocol/Makefile
+++ b/server/modules/protocol/Makefile
@@ -31,7 +31,7 @@ UTILSPATH := $(ROOT_PATH)/utils
 
 CC=cc
 CFLAGS=-c -fPIC -I/usr/include -I../include -I../../include -I$(LOGPATH) \
-	-I$(UTILSPATH) -Wall -g
+	-I$(UTILSPATH) -Wall -g -fprofile-arcs -ftest-coverage
 
 include ../../../makefile.inc
 
@@ -51,7 +51,7 @@ MAXSCALEDOBJ=$(MAXSCALEDSRCS:.c=.o)
 SRCS=$(MYSQLCLIENTSRCS) $(MYSQLBACKENDSRCS) $(TELNETDSRCS) $(HTTPDSRCS) \
 	$(MAXSCALEDSRCS)
 OBJ=$(SRCS:.c=.o)
-LIBS=$(UTILSPATH)/skygw_utils.o 
+LIBS=$(UTILSPATH)/skygw_utils.o -lgcov
 MODULES=libMySQLClient.so libMySQLBackend.so libtelnetd.so libHTTPD.so \
 	libmaxscaled.so
 
diff --git a/server/modules/routing/Makefile b/server/modules/routing/Makefile
index dbb0503..8304766 100644
--- a/server/modules/routing/Makefile
+++ b/server/modules/routing/Makefile
@@ -29,7 +29,7 @@ UTILSPATH := $(ROOT_PATH)/utils
 
 CC=cc
 CFLAGS=-c -fPIC -I/usr/include -I../include -I../../include -I$(LOGPATH) \
-	-I$(UTILSPATH) -Wall -g
+	-I$(UTILSPATH) -Wall -g -fprofile-arcs -ftest-coverage
 
 include ../../../makefile.inc
 
@@ -48,7 +48,7 @@ WEBSRCS=webserver.o
 WEBOBJ=$(WEBSRCS:.c=.o)
 SRCS=$(TESTSRCS) $(READCONSRCS) $(DEBUGCLISRCS) cli.c
 OBJ=$(SRCS:.c=.o)
-LIBS=$(UTILSPATH)/skygw_utils.o -lssl -llog_manager
+LIBS=$(UTILSPATH)/skygw_utils.o -lssl -llog_manager -lgcov
 MODULES= libdebugcli.so libreadconnroute.so libtestroute.so libcli.so \
 	libwebserver.so
 
diff --git a/server/modules/routing/readwritesplit/Makefile b/server/modules/routing/readwritesplit/Makefile
index c60f2ff..a3a643e 100644
--- a/server/modules/routing/readwritesplit/Makefile
+++ b/server/modules/routing/readwritesplit/Makefile
@@ -27,7 +27,7 @@ QCLASSPATH 	:= $(ROOT_PATH)/query_classifier
 CC=cc
 CFLAGS=-c -fPIC -I/usr/include -I../../include -I../../../include \
 	-I$(LOGPATH) -I$(UTILSPATH) -I$(QCLASSPATH) \
-	$(MYSQL_HEADERS) -Wall -g
+	$(MYSQL_HEADERS) -Wall -g -fprofile-arcs -ftest-coverage
 
 include ../../../../makefile.inc
 
@@ -38,7 +38,7 @@ LDFLAGS=-shared -L$(LOGPATH) -L$(QCLASSPATH) -L$(EMBEDDED_LIB) \
 
 SRCS=readwritesplit.c
 OBJ=$(SRCS:.c=.o)
-LIBS=-lssl -pthread -llog_manager -lquery_classifier -lmysqld
+LIBS=-lssl -pthread -llog_manager -lquery_classifier -lmysqld -lgcov
 MODULES=libreadwritesplit.so
 
 all:	$(MODULES)
